### ISSUE TITLE
Unable to specify a default value for a generic parameter

### ISSUE THREAD

**Issue by rowillia** [python/mypy#3737]
Simplified Example:

```python
from typing import TypeVar

_T = TypeVar('_T')
def foo(a: _T = 42) -> _T:  # E: Incompatible types in assignment (expression has type "int", variable has type "_T")
    return a
```

Real World example is something closer to:

```python
_T = TypeVar('_T')
def noop_parser(x: str) -> str:
    return x

def foo(value: str, parser: Callable[[str], _T] = noop_parser) -> _T:
    return parser(value)
```

**Comment by gvanrossum** [comment]
I think the problem here is that *in general*, e.g. if there are other parameters also using `_T` in their type, the default value won't work. Since this is the only parameter, you could say that this is overly restrictive, and if there are no parameters, the return type should just be determined by that default value (in your simplified example, you'd want it to return `int`). But I'm not so keen on doing that, since it breaks as soon as another generic parameter is used.

To make this work, you can use `@overload`, e.g.
```py
@overload
def foo() -> int: ...
@overload
def foo(a: _T) -> _T: ...
def foo(a = 42):  # Implementation, assuming this isn't a stub
    return a
```

That's what we do in a few places in typeshed too (without the implementation though), e.g. check out `Mapping.get()` in [typing.pyi](https://github.com/python/typeshed/blob/master/stdlib/3/typing.pyi#L309).

**Comment by OddBloke** [comment]
I'm running in to a similar issue when trying to iron out a bug in the configparser stubs. Reducing RawConfigParser to the relevant parts gives:

```py
_section = Mapping[str, str]

class RawConfigParser:
    def __init__(self, dict_type: Mapping[str, str] = ...) -> None: ...

    def defaults(self) -> _section: ...
```

This is incorrect at the moment, because `dict_type` should (as the name indicates) be a type. Furthermore, `defaults()` returns an instance of `dict_type`, so this seems like a good application of generics. The following works as expected when a `dict_type` is passed (i.e. `reveal_type(RawConfigParser(dict_type=dict).default())` gives `builtins.dict*[Any, Any]`):

```py
_T1 = TypeVar('_T1', bound=Mapping)

class RawConfigParser(Generic[_T1]):
    def __init__(self, dict_type: Type[_T1] = ...) -> None: ...

    def defaults(self) -> _T1: ...
```

but when you don't pass a dict_type, mypy can't infer what type it should be using and asks for a type annotation where it is being used. This is less than ideal for the default instantiation of the class (which will be what the overwhelming majority of people are using). So I tried adding the appropriate default to the definition:

```py
_T1 = TypeVar('_T1', bound=Mapping)

class RawConfigParser(_parser, Generic[_T1]):
    def __init__(self, dict_type: Type[_T1] = OrderedDict) -> None: ...

    def defaults(self) -> _T1: ...
```

but then I hit the error that OP was seeing:

```
stdlib/3/configparser.pyi:51: error: Incompatible default for argument "dict_type" (default has type Type[OrderedDict[Any, Any]], argument has type Type[_T1])
```

This feels like something that isn't currently supported, but I'm not sure if I'm missing a way that I could do this...

**Comment by ilevkivskyi** [comment]
@OddBloke 
This default is, strictly speaking, not safe. This would require a _lower_ bound for ``_T1`` in order to be safe.

**Comment by OddBloke** [comment]
@ilevkivskyi I don't follow, I'm afraid; could you expand on that a little, please?

**Comment by ilevkivskyi** [comment]
Consider this code:
```python
class UserMapping(Mapping):
    ...
RawConfigParser[UserMapping]()
```
In this case ``OrderedDict`` will be "assigned" to ``dict_type``, which will be ``Type[UserMapping]``, which is not a supertype of ``Type[OrderedDict]``. In order for the default value to be safe it is necessary to have some guarantee that ``_T1`` will be always a _supertype_ of ``OrderedDict`` (this is called a lower bound), but this is not supported, only upper bounds are allowed for type variables.

**Comment by OddBloke** [comment]
@ilevkivskyi  Thanks, that makes sense. Do you think what I'm trying to do is unrepresentable as things stand?

**Comment by ilevkivskyi** [comment]
I didn't think enough about this, but my general rule is not to be "obsessed" with precise types, sometimes ``Any`` is OK. If a certain problem appears repeatedly, then maybe we need to improve something (also now we have plugin system). In this particular case my naive guess would be to play with overloads, as Guido suggested (note that ``__init__`` can also be overloaded).

**Comment by nematsakis** [comment]
Allowing default values for generic parameters might also enable using type constraints to ensure that generics are only used in specified ways.  This is a bit esoteric, but here is an example of how you might use the same implementation for a dict and a set while requiring that consumers of the concrete dict always provide a value to `insert()` while consumers of the `set()` type never do.

```
class Never(Enum):
    """An uninhabited type"""

class Only(Enum):
    """A type with one member (kind of like NoneType)"""
    ONE = 1

NOT_PASSED = Only(1)

class MyGenericMap(Generic[K, V]):

    def insert(key, value=cast(Never, NOT_PASSED)):
        # type: (K, V) -> None
        ...
       if value is NOT_PASSED:
           ....

class MySet(MyGenericMap[K, Never]):
    pass

class MyDict(MyGenericMap[K, V]):
    pass

my_set = MySet[str]()
my_set.insert('hi')           # this is fine, default value matches concrete type
my_set.insert('hi', 'hello')  # type error, because 'hello' is not type never

my_dict = MyDict[str, int]()
my_dict.insert('hi')          # type error because value is type str, not Never
my_dict.insert('hi', 22)      # this is fine, because the default value is not used
```

**Comment by gvanrossum** [comment]
Huh, this just came up in our code.

**Comment by ilevkivskyi** [comment]
This request already appeared five times, so I am raising priority to high.

I think my preferred solution for this would be to support lower bounds for type variables. It is not too hard to implement (but still a large addition), and at the same time it may provide more expressiveness in other situations.

**Comment by hynekcer** [comment]
A second part for the workaround with overload should be noted: The header of the overloaded function must be without annotation, but it is not acceptable for more complex functions that of course should be checked. Only the binding between input and output types is checked by overloading. The body can be encapsulated by a new function with the same types, without a default.

```python
@overload
def foo() -> int: ...
@overload
def foo(a: _T) -> _T: ...

def foo(a = 42):  # unchecked implementation
    return foo_internal(a)

def foo_internal(a: _T) -> _T:
    # a checked complicated body moved here
    return a
```
Everything seems that can be checked even in much more complicated cases, even in much more complicated cases, but the number of necessary declarations with @overload can rise exponentially by 2 ** number_of generic_vars_with_defaults.

**Comment by hynekcer** [comment]
Another solution is to use a broad static type for that parameter and immediately assign it in the body to the original exact generic type. It is a preferable solution for a class overload.  
An example is a database cursor with a generic row type:
``` python
_T = TypeVar('_T', list, dict)

class Cursor(Generic[_T]):
    @overload
    def __init__(self, connection: Connection) -> None: ...
    @overload
    def __init__(self, connection: Connection, row_type: Type[_T]) -> None: ...
    def __init__(self, connection: Connection, row_type: Type=list) -> None:
        self.row_type: Type[_T] = row_type  # this annotation is important
        ...
    def fetchone(self) -> Optional[_T]: ...
    def fetchall(self) -> List[_T]: ...   # more methods depend on _T type

cursor = Cursor(connection, dict)  # cursor.execute(...)
reveal_type(cursor.fetchone())     # dict
```

**Comment by andersk** [comment]
> I think the problem here is that *in general*, e.g. if there are other parameters also using `_T` in their type, the default value won't work.

Wait. Why is that a problem? If default value doesn’t work for some calls, that should be a type error *at the call site*.

```python
_T = TypeVar('_T')
def foo(a: _T = 42, b: _T = 42) -> List[_T]:
    return [a, b]

foo()  # This should work.
foo(a=17)  # This should work.
foo(b=17)  # This should work.
foo(a="foo")  # This should be a type error (_T cannot be both str and int).
foo(b="foo")  # This should be a type error (_T cannot be both int and str).
foo(a="foo", b="foo")  # This should work.
```

> This default is, strictly speaking, not safe. This would require a *lower* bound for _T1 in order to be safe.

Only if we retain the requirement that the default value always works, which is not what’s being requested here. The desired result is:

```python
_T1 = TypeVar('_T1', bound=Mapping)
class RawConfigParser(_parser, Generic[_T1]):
    def __init__(self, dict_type: Type[_T1] = OrderedDict) -> None: ...

    def defaults(self) -> _T1: ...

class UserMapping(Mapping):
    ...

RawConfigParser[OrderedDict]()  # This should work.
RawConfigParser[UserMapping]()  # This should be a type error (_T1 cannot be both UserMapping and OrderedDict).
RawConfigParser[UserMapping](UserMapping)  # This should work.
```

**Comment by radeksz** [comment]
I ran into this trying to write
```python
_T = TypeVar('_T')
def str2int(s:str, default:_T = None) -> Union[int, _T]: # error: Incompatible default for argument "default" (default has type "None", argument has type "_T")
    try:
        return int(s)
    except ValueError:
        return default
```
(after figuring out that I need `--no-implicit-optional` mypy parameter as well)

It all raises the bar for type-hinting newbies.

**Comment by hauntsaninja** [comment]
This comes up a lot (and will likely come up even more if we ever get around to making `--no-implicit-optional` the default), so I'd like to spew some thoughts in the hope that it leads to something.

I believe what Anders is proposing is:
1) When checking a definition ignore the default value, but keep it around for later
2) Run mypy's existing type inference at call sites (crucially, continuing to ignore the default value)
3) Check that default value's type is a subtype of whatever we solved for T (assuming `x: T = ...`, but doing the appropriate thing for `x: Optional[T] = ...` and so on)

This is different from what I believe Ivan is proposing in that it would cause foo to behave unlike `bar` in the following:
```
def foo(a: T = 42, b: T = 42) -> List[T]: ...
def bar(a: T, b: T) -> List[T]: ...

reveal_type(foo("asdf"))  # type error for Anders, because default value 42 is not a str
reveal_type(bar("asdf", 42)) # reveals List[object]
```
Or if you're the kind of person who hates joins that end with object, maybe you'll find this more intuitive:
```
class X: ...
class Y(X): ...

def foo(a: T = X(), b: T = X()) -> List[T]: ...
def bar(a: T, b: T) -> List[T]: ...

reveal_type(foo(Y()))  # type error for Anders, because default value X() is not a Y
reveal_type(bar(Y(), X())) # reveals List[X]
```

I think typevar lower bounds would result in behaviour generally more consistent with what mypy does elsewhere. Since Anders' proposal is less permissive than type var lower bounds, it's also more likely to be disruptive. But the differences here might be somewhat moot, since I reckon the implementations would have a good bit in common: for starters, both will have to modify parsing and CallableType to keep track of default values. Type var lower bounds would need to do some subtraction to associate a CallableType.variable in cases like `Optional[T]` (or worse, `List[T]`), but Anders' would have to do something similar in applytype. Typevar lower bounds would probably have to touch more code, though.

A fun side note about both solutions is causing default values to matter to type checking might mean we'd sometimes want to use them in stubs. Although with lower bounds, we could add a lower bound attribute to typing.TypeVar which stubs could use (instead of inferring it from default values)...

**Comment by andersk** [comment]
> I believe what Anders is proposing is:
> 
> 1. When checking a definition ignore the default value, but keep it around for later
> 2. Run mypy's existing type inference at call sites (crucially, continuing to ignore the default value)

No, at the call site we know whether the default value is being used, so we don’t need to ignore it; we can allow it to participate in type inference, *if it is being used*.

To put it another way, we treat

```python
def foo(a: T = X(), b: T = X()) -> List[T]: ...
foo(Y())
```

as syntactic sugar for

```python
a_default = X()
b_default = X()
def foo(a: T, b: T) -> List[T]: ...
foo(Y(), b_default)
```

**Comment by hauntsaninja** [comment]
Thanks for clarifying! Just to make sure I get things straight, this means the `foo(a="foo")` and `foo(b="foo")` examples in your original comment wouldn't in fact be a type error, since `T = object` is a totally valid inference.
In which case I'm not sure I see a difference in semantics between lower bounded typevars and what you're propsing; it's more just a difference in implementation (in one case, the lower bound is something we automatically add to the TypeVarDef, in the other the lower bound is an additional constraint we infer from the CallableType).

**Comment by andersk** [comment]
Yes, strictly speaking, you’re right about `_T = object` in the [first example](https://github.com/python/mypy/issues/3737#issuecomment-552840010). I’ve edited it to use invariant `List[_T]` types to avoid that possibility. The second example stands as written.

Lower bounded typevars do not solve the same problem; they don’t allow you to provide defaults unless the defaults would be valid at all call sites, even call sites where the default is not used (e.g. `RawConfigParser[UserMapping](UserMapping)`).

**Comment by efagerberg** [comment]
Seems like the current workaround is to wrap the parameter in an optional and handle the default in code.

Is there any update on this in general? These kind of things do make it harder to get larger adoption.

**Comment by andersk** [comment]
@efagerberg `Optional` is not a workaround at all, neither for the caller side nor for the callee side.

```python
from typing import Optional, TypeVar

_T = TypeVar("_T")
def foo(a: Optional[_T] = None) -> _T:
    if a is None:
        return 42  # error: Incompatible return value type (got "int", expected "_T")
    return a

b = foo()  # error: Need type annotation for 'b'

c: str = foo()  # false negative!
c.lower()  # AttributeError: 'int' object has no attribute 'lower'
```

Overloads sort of help, but they are not checked on the callee side, and have the exponential explosion issue explained [above](https://github.com/python/mypy/issues/3737#issuecomment-465355737).

**Comment by tilsche** [comment]
@hynekcer I cannot get the suggested workaround for class generic types running unless the `TypeVar` is limited to concrete types. Otherwise, as soon the the actual default parameter is used (i.e. drop the `dict` from construction), mypy will complain `Need type annotation for "cursor"` and reveal an `Optional[Any]` rather than `Optional[list]`. When I just change the order of types in the `TypeVar`, it break and I get the *wrong* type revealed.

I must say, among the many small annoyances this is by far the biggest typing issue I encountered so far. I need to either compromise on actual non-typing functionality, i.e. no default arguments or vastly different interface, or completely loose the generic typing for the class. You can't just slap a `type: ignore` on some small hidden statement and there seems to be no viable workaround.

**Comment by socketpair** [comment]
Consider the following:
```python
from typing import TypeVar, Generic, overload, Type, Optional, List, TypedDict, Any, cast

T = TypeVar('T',  str, int)

class BaseUpdater(Generic[T]):
    @overload
    def __init__(self) -> None: ...

    @overload
    def __init__(self, rowType: Type[T]) -> None: ...

    def __init__(self, rowType: Type=str) -> None:
        self.rowType: Type[T] = rowType  # this annotation is important
        pass

    def fun(self) -> T:
        return cast(T, self.rowType('42'))

class Updater(BaseUpdater[T]):
    pass


reveal_type(Updater(int).fun())  # int
reveal_type(Updater(str).fun()) # str

# depends on order str|int or int|str in T= ..... line !!! WHY ?! "Type=str" or "Type=int" is irrelevant in constructor (!)
reveal_type(Updater().fun())   

```

`T = TypeVar('T',  bound=str|int)` make this scheme not working. Last reveal_type will show `Revealed type is "<nothing>"`

**Comment by markedwards** [comment]
Coming up on the fifth anniversary of this issue.  Shouldn't this either be addressed or classified as "won't fix"?  This limitation makes it overly complicated to annotate what should be a fairly basic case, so it would be good to know if it will ever change.

**Comment by NeilGirdhar** [comment]
@markedwards Default values were recently drafted in [PEP 696](https://peps.python.org/pep-0696/) slated for Python 3.12, and have been added to [typing-extensions](https://github.com/python/typing_extensions) already.  Either this issue could be renamed to track its implementation or you could create one.  The MyPy team has been doing amazing work with [Python 3.11 issues](https://github.com/python/mypy/issues/12840).  I am just guessing, but I imagine they will start on the Python 3.12 issues after the 3.11 issues.

**Comment by markedwards** [comment]
@NeilGirdhar, fair enough.  So then the simplified example in the original issue would become:

```
from typing_extensions import TypeVar

_T = TypeVar('_T', default=int)
def foo(a: _T = 42) -> _T:
    return a
```

Not confirmable yet: https://mypy-play.net/?mypy=master&python=3.12&flags=strict&gist=d083d1bbf2f8f9426f876866947834a0

**Comment by NeilGirdhar** [comment]
@markedwards Looks right to me.  Yes,  as far as I know, no one implements PEP 696 yet.

**Comment by AlexWaygood** [comment]
> as far as I know, no one implements PEP 696 yet.

Pyright has implemented PEP 696.

**Comment by NeilGirdhar** [comment]
> Pyright has implemented PEP 696.

Wow!  Thanks for letting me know 😄

**Comment by andersk** [comment]
PEP 696 is different from this issue.

PEP 696 is about providing a **type-level default** for a type variable used in a generic type—example from the PEP:

```python
T = TypeVar("T", default=int)

@dataclass
class Box(Generic[T]):
    value: T | None = None
```

This issue is about providing a **value-level default** for a function parameter whose type is generic—example from the description:

```python
_T = TypeVar('_T')
def foo(a: _T = 42) -> _T:
    return a
```

**Comment by markedwards** [comment]
To @andersk's point, I also don't see how PEP 696 makes any difference here.  The `TypeVar` default does not address the core issue here, which is that the default value might conflict with the type variable.

For what its worth, Typescript also does not support this pattern:
<img width="962" alt="Screenshot 2023-02-28 at 08 34 29" src="https://user-images.githubusercontent.com/454900/221806033-ac089520-948d-4205-864c-c047eef3c78d.png">

And adding a type variable default in Typescript does not "fix" the issue:
<img width="1024" alt="Screenshot 2023-02-28 at 08 35 24" src="https://user-images.githubusercontent.com/454900/221806241-7cf22ddf-6432-4659-ba90-317045959648.png">

That's not to say that Python typing is for any reason bound by what Typescript can do, just that this limitation is not a peculiarity of Python or mypy.

It seems to me that a final decision should be made here and this should be documented so developers don't need to find this thread to sort it out.

**Comment by NeilGirdhar** [comment]
> To @andersk's point, I also don't see how PEP 696 makes any difference here. 

Sorry, that was my mistake. I hadn't looked at this issue carefully since I subscribed to it, and forgot what it was actually about.

**Comment by JukkaL** [comment]
I'm currently doing some experimentation in the hope of eventually addressing this.

**Comment by JukkaL** [comment]
I haven't made much progress with an implementation yet, but here are some notes based on what I've learned so far (much of this is also covered above in the discussion, but I want so summarize everything in one comment):
* Storing the types of default values in a callable type and using these to generate additional constraints when default arguments aren't provided at a call site seems like the most promising approach. We'd only store the default value types for generic functions that have default arguments to minimize the impact to memory use and performance.
* We can't infer the types of default values during semantic analysis. This means that we can't always produce the signature of a function during semantic analysis any more. A possible workaround is to infer the signature of a generic callable that accepts defaults lazily during type checking, similar to what we do with decorated functions. We probably don't want to do this for all functions, since it could be quite a disruptive change.
* We need to make the check for valid default argument values in a function definition more lenient.
* For arguments such as `x: T | None = None`, we wouldn't generate a constraint since the value of the default value is also covered by the argument type. These cases may have additional trickiness as well.

**Comment by ilevkivskyi** [comment]
Yeah, this is essentially a plan I had in mind a while ago. But after all it seemed to me this would cause a visible performance impact even with all the possible optimizations. OTOH it may be fine with all the recent performance improvements. Btw few more points to keep in mind:
* We may need to special-case typeshed stubs (and maybe protocols) in cases like `def foo(x: T = ...) -> T: ...` (with literal `...` as default). Also this change will mean providing actual default will change the type in stubs, i.e. `def foo(x: T = ...) -> T: ...` and `def foo(x: T = 0) -> T: ...` would be two different types.
* All the type ops for callables (and other types, e.g. overloads) like is_subtype, meet, join, etc., will need to be reconsidered to account for default types.
* There may be some trickiness if default values are lambdas, since there is a very old bug when applying generic function to a generic function, it may cause troubles here.

**Comment by tpvasconcelos** [comment]
Just ran into this one as well... Is the temporary "solution" most of you are going with to just place a `# type: ignore[assignment]` comment until this issue gets resolved?

i.e., using the example from @OddBloke and @andersk's :
```python
_T1 = TypeVar('_T1', bound=Mapping)
class RawConfigParser(_parser, Generic[_T1]):
    def __init__(
        self, 
        dict_type: Type[_T1] = OrderedDict,  # type: ignore[assignment]
    ) -> None: ...

    def defaults(self) -> _T1: ...
```

**Comment by couling** [comment]
> A second part for the workaround with overload should be noted: The header of implementation of the overloaded function must be without a generic annotation for that variable, but it is not acceptable for more complex functions where the body should be checked. Only the binding between input and output types is checked by overloading.

May I ask if there's a little more detail on this design decision?  Why are funciton implementations not checked for overloads?

Searches on the topic are giving me blanks.  I'm curious why mypy would not check a defanition for each overload variant.  It would only need to revist one function per overload so I can't see the cost would be that high, nor the implementation that complex.  I wonder what I'm missing?

**Comment by bswck** [comment]
Let's close this due to [PEP 696](https://peps.python.org/pep-0696/)?
@JelleZijlstra

**Comment by hauntsaninja** [comment]
I don't think this should be closed. I don't think this should require specifying a default on the TypeVar explicitly and also mypy's behaviour doesn't change if you do so (default type is little different from default value)

**Comment by bswck** [comment]
@hauntsaninja In that case--sorry for causing the mess.

**Comment by mcmanustfj** [comment]
I'm looking at this and it seems like one potential solution would be updating `TypeChecker.check_default_args` to replace the types of generic arguments with the types of their initializers - i.e `def f[T](var: T = "asdf") -> T: ...` is actually checked as `def f(var: str = "asdf") > str`, but I'm not sure how to infer `str` and not `Literal["asdf"]` from the context of a type checker.

**Comment by andersk** [comment]
@mcmanustfj That is not a solution because we do not want the default’s type to constrain call sites where the default is not used. In your example, we need to accept `f(42)` returning `int`. See my [comment above](https://github.com/python/mypy/issues/3737#issuecomment-552840010) for details of what should be accepted and rejected.

**Comment by mcmanustfj** [comment]
@andersk My thought was to only change the function type checking and not the invocation type checking, but I could definitely be misunderstanding how check_default_args works.

My test patch has the following change: 

```
    def check_default_args(self, item: FuncItem, body_is_trivial: bool) -> None:
        if any(
            arg.initializer is not None and type(arg.type_annotation) is UnboundType
            for arg in item.arguments
        ):
            args_without_generic_defaults = [
                Argument(
                    variable=arg.variable,
                    type_annotation=arg.type_annotation,
                    initializer=None
                    if type(arg.type_annotation) is UnboundType
                    else arg.initializer,
                    kind=arg.kind,
                    pos_only=arg.pos_only,
                )
                for arg in item.arguments
            ]

            item_without_generic_defaults: FuncItem = FuncItem(
                arguments=args_without_generic_defaults,
                body=item.body,
                typ=item.type,
                type_args=item.type_args,
            )

            self.check_default_args(item_without_generic_defaults, body_is_trivial)
            return
    ... # rest of check_default_args
```

Right now, 
```
def test_func(var: T = "asdf") -> T:
    return var

if __name__ == "__main__":
    x = test_func()
    reveal_type(x)
    reveal_type(test_func())
    reveal_type(test_func("asdf"))
    reveal_type(test_func(True))
    reveal_type(test_func(1))
```
has the opposite problem and doesn't give the default invocation a type at all, 
```
example.py:31: note: Revealed type is "Any"
example.py:32: note: Revealed type is "Never"
example.py:33: note: Revealed type is "builtins.str"
example.py:34: note: Revealed type is "builtins.bool"
example.py:35: note: Revealed type is "builtins.int"
```

Not sure where to go from here, unfortunately.

**Comment by andersk** [comment]
@mcmanustfj Just conceptually, changing only the function type checking as you describe is still not a solution, because it would fail to reject this wrong definition:

```python
def f[T](var: T = "asdf") -> T:
    # we should not be able to assume var: str here
    return var.lower()

f(42)
```

and fail to reject this wrong invocation:

```python
def f[T](var: T = "asdf") -> T:
    return var

ret: int = f()
```

The correct solution needs to involve type-checking the default value *at each call site* if the default is used there. That is, we need to treat

```python
def f[T](var: T = "asdf") -> T:
    return var

f()
f(arg)
```

as the syntactic sugar it is for something like

```python
f_var_default = "asdf"
def f[T](var: T) -> T:
    return var

f(f_var_default)
f(arg)
```

(with perhaps an additional definition-site check that there _exists_ some `T` that unifies the types of `f_var_default` and `var`, but that’s not required for soundness, as the consequence if there doesn’t would be that a call relying on the default is safely prohibited).

**Comment by MestreLion** [comment]
Hi there! mypy shows a type error in my code which Google's AI vaguely said is a mypy limitation (without citing any source), digging further I've found this github issue which I _think_ is the right one. So, 3 questions:

1 - Is my error a manifestation of the issue discussed here? (i.e, am I posting in the right place?)

```python
# test.py

def csv2list[T](text:str, sep:str=",", itemtype:type[T]=str) ->  list[T]:
    return [itemtype(_.strip()) for _ in (text.split(sep) if text else [])]

print(csv2list("a, b, c"))
print(csv2list("1, 2, 3", itemtype=int))
```
```console
$ python3.12 test.py
['a', 'b', 'c']
[1, 2, 3]
$ mypy test.py
test.py:3: error: Incompatible default for argument "itemtype" (default has type "type[str]", argument has type "type[T]")  assignment]
test.py:4: error: Too many arguments for "object"  [call-arg]
Found 2 errors in 1 file (checked 1 source file)
$ mypy --version
mypy 1.17.1 (compiled: yes)
```

2 - Is this the correct way of annotating my function, and the error is a known bug/limitation in mypy?

3 - Are there any workarounds besides `@overload`? Would using pre-3.12 syntax of `T = typing.TypeVar("T")` help?

**Comment by tmke8** [comment]
Re 1, 2: Yes. You can tell it's a mypy limitation because it works in pyright, which you can try out here: https://pyright-play.net/

Though you should annotate `itemtype` as `Callable[[str], T]` instead of `type[T]` because you don't just require a type -- you require a constructor that takes a `str` as an argument. (This doesn't help you with the mypy limitation though.)

So, I would write it [like this](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDCAhgDYlEBGJApgFC0Am1wUAxgM4BuATCUuzADaAFQC6AChjUAHjABcAkABoo7aggUwQAXgBES-ZikR4CanMKlyVaoMGLRKsdsUBKKAFoAfFCh8BIqJytL6%2BINQwAK4gKFCCSMam1OIA%2BgB0ikgI4q7uwLhQKZixkjIwGQh8MOJqCO5ILFKyUNQkanGirqL0CCCo1Rw8-tW6RCoUKqy6ubS9-eKDvPwjAIwq3CoAzIYJ1CaI1Nr9uUA).

3: Pre-3.12 syntax will not help. I think what I would do with your code is disentangling the two functionalities of the function: have one function that just does the splitting and stripping; and have another that does conversion if needed.

**Comment by MestreLion** [comment]
@tmke8 , thank you so much! I actually began with `Callable`, then trimmed down to `type[]` trying to please mypy.

Regarding your suggestion of splitting the function, it would simply push the bug down to the "conversion" function, no? `csv2list()` would get rid of  the problematic `itemtype` and return `list[str]`, becoming a non-generic function. But if I understood the bug correctly, the new function would have the same signature regarding `itemtype`, triggering the (buggy) behavior.

**Comment by tmke8** [comment]
I would probably write it like this:

```python
def csv2list(text:str, sep:str=",") -> list[str]:
    return [_.strip() for _ in (text.split(sep) if text else [])]

print(csv2list("a, b, c"))
print(map(int, csv2list("1, 2, 3")))
```

**Comment by RomanValov** [comment]
bumped into the same issue with following snippet:

```python
def func[T](callback: Callable[[], T] = lambda: None) -> T:
    return callback()
```

mypy raises two errors:

```
error: Incompatible default for parameter "callback" (default has type "Callable[[], None]", parameter has type "Callable[[], T]")  [assignment]
error: Incompatible return value type (got "None", expected "T")  [return-value]
```

not sure though if second error (return-value) is a consequence of failed binding of default value and generic parameter.
